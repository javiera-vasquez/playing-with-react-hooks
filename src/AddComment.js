import React, { useState } from "react";

const AddComment = ({ id, cb, questionId, creatorId }) => {
  const [inputValue, setImputValue] = useState('');
  return (
    <form onSubmit={e => {
      cb({
        questionId,
        value: {
          comment: inputValue, createdAt: new Date(), id, creatorId
        }
      });
      setImputValue('');
      e.preventDefault();
    }}>
      <input
        type="text"
        value={inputValue}
        onChange={e => setImputValue(e.target.value)}
        className="add-comment-input"
        placeholder="Add commentâ€¦"
      />
    </form>
  )
}

export default AddComment;