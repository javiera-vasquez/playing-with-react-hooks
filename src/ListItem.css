.list-item {
  padding: 10px;
  font-size: 15px;
  font-weight: 500;
  line-height: 1.6
}

.list-item.question {
  position: relative;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.list-item.question::after {
  right: 10px;
  top: 20px;
  content: "\003e";
  position: absolute;
  transform: rotate(90deg);
  transition: all .2s ease-out;
}

.list-item.question.active::after {
  transform: rotate(-90deg);
}

.list-item.awnser {
  display: none;
  background-color: rgba(0, 0, 0, 0.01);
}

.list-item.awnser.active {
  display: block;
}

.list-item .list-item__user-creation {
  color: #869596;
  font-size: 14px;
}

.list-item .list-item__user-creation strong {
  margin-right: 10px;
}