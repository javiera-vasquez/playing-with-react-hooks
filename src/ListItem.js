import React from "react";
import "./ListItem.css";

const ListItem = ({...props}) => {
  const { type, user, content, date } = props;
  return (
    <div class={`list-item ${type}`}>
      <div class="list-item__user-creation">
        <strong>{user}</strong>{date && <span>{date}</span>}
      </div>
      <div class="list-item__content">{content}</div>
    </div>
  );
};

export default ListItem;
