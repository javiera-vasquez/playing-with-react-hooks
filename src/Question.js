import React from "react";
import { map, reduce, find } from "lodash";

import usersRespose from "./api/users";
import ListItem from "./ListItem";
import AddComment from "./AddComment";

import { setDateFormat, dateEventsToTrack } from "./helpers";

const Question = ({
  question, answer, isActive, onClickCb, activeUser, onAddCommentCb
}) => {
  const getlastEdit = (list, events = dateEventsToTrack) => {
    return setDateFormat(
      reduce(events, (acc, e) => list[e] !== null ? list[e] : acc, undefined)
    )
  }

  return (
    <div
      className={`question-wrapper ${isActive ? 'active' : ''}`}
      onClick={() => onClickCb(question.id)}
    >
      <ListItem
        type={`question`}
        active={isActive}
        user={
          activeUser.organization === question.user.organization ?
            question.user.name : question.user.organization
        }
        content={question.questionText}
        date={getlastEdit(question)}
      />
      { answer &&
        <ListItem
          type={`awnser`}
          active={isActive}
          user={
            activeUser.organization === answer.user.organization ?
            answer.user.name : answer.user.organization
          }
          content={answer.answerText}
          date={getlastEdit(answer)}
          child={
            <div className="list-item__comments-wrapper">
              <AddComment
                id={question.comments.length + 1}
                cb={comment => onAddCommentCb(comment)}
                questionId={question.id}
                creatorId={activeUser.id}
              />
              { !!question.comments.length &&
                map(question.comments, comment => {
                  const user = find(usersRespose, {id: comment.creatorId});
                  if(user.organization !== activeUser.organization) return null;
                  return (
                    <div className="list-item__comments" key={comment.id}>
                    <strong>{user.name}:</strong>
                    <span>{comment.comment}</span>
                    <div className="list-item__comments__date">
                      {setDateFormat(comment.createdAt)}
                    </div>
                  </div>
                  )
                })
              }
            </div>
          }
        />
      }
    </div>
  )
}

export default Question;