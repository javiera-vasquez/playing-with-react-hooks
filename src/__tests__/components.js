import React, { useState } from 'react';
import TestRenderer from 'react-test-renderer';

import questionsResponse from "../api/questions";
import answersRespose from "../api/answers";
import usersRespose from "../api/users";

import Question from '../Question';
import ListItem from "../ListItem";
import AddComment from "../AddComment";

describe('Init Question component', () => {
  it('Need to render()', () => {
    const component = TestRenderer.create(
      <Question
        question={{
          ...questionsResponse[0],
          user: usersRespose[0]
        }}
        answer={false}
        isActive={true}
        onClickCb={() => {}}
        activeUser={usersRespose[0]}
        onAddCommentCb={() => {}}
      />
    ).toJSON();
    expect(component).toMatchSnapshot();
  })
})

describe('Init ListItem component', () => {
  it('It render()', () => {
    const component = TestRenderer.create(
      <ListItem
        type="test"
        user="test"
        content="hola soy un test"
        date={"Oct 28, 4:27 PM"}
        active={true}
        child={<span>hola</span>}
      />
    ).toJSON();

  })
})

// Need more research
// React create app can't not handle hooks component in jest env yet
// describe('Init AddCommnet component', () => {
//   it('It render()', () => {
//     const component = TestRenderer.create(
//       <AddComment
//         id={1} cb={() =>{}} questionId={1} creatorId={1}
//       />
//     ).toJSON();
//     expect(component).toMatchSnapshot();
//   })
// })
