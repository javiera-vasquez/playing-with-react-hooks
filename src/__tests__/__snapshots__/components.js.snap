// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init Question component Need to render() 1`] = `
<div
  className="question-wrapper active"
  onClick={[Function]}
>
  <div
    className="list-item question active"
  >
    <div
      className="list-item__user-creation"
    >
      <strong>
        Alonso Armani
      </strong>
      <span>
        Oct 28, 4:27 PM
      </span>
    </div>
    <div
      className="list-item__content"
    >
      What was the name of Danny's gang in 'Grease'?
    </div>
  </div>
</div>
`;
